paragraph/scheduling/graph_scheduler.h

```diff
@@ -42,7 +43,7 @@ class GraphScheduler {
   // scheduled.
   // Part of Public API with simulators
   static shim::StatusOr<std::unique_ptr<GraphScheduler>> Create(
-      Graph* graph);
+      Graph* graph, const std::string& filename = "");
```

Comment:
What if instead of giving the string to the scheduler, you give it the unique_ptr to a logger (better yet, make a SetLogger() method). Then you can drop the IsAvailable function call altogether and just check the logger for nullptr. All Logger objects would actually be logging.

In libdes (my parallel sim framework) I have a vector of logger pointers. All log events go to all logger objects. That way the user can create multiple loggers. For example, they could make a logger with filters that only look at send/recv, and another logger that looks at collectives. I think I defined a vector of raw pointers and implemented an AddLogger() method.

Reply:
I was thinking exactly about that, but then thought that it's not simple enough, so I decided to make a single logger per scheduler, for now. In that case, unique_ptr looks like an okay choice to make sure the logger is unique :)
You can add issue/feature_request for multiple loggers with filters, and we can add them later. For now I exposed interface to set the logger from scheduler, but didn't expose logger itself as I didn't see need.